File: 0030. Substring with Concatenation of All Words.java
// Hard 
// Sliding Window, Hash Table
// Time:O(n*wordLen),Space:O(m)
// https://leetcode.cn/problems/substring-with-concatenation-of-all-words/

import java.util.*;
class Solution {
    public List<Integer> findSubstring(String s, String[] words) {
        List<Integer> res = new ArrayList<>();
        // if (s == null || s.length() == 0 || words == null || words.length == 0) return res;
        int wordLen = words[0].length(), numWord = words.length, windowSize = wordLen * numWord;

        // use Map to store the appear time of each char
        Map<String, Integer> map = new HashMap<>();
        for (String word : words) {
            map.put(word, map.getOrDefault(word, 0) + 1);
        }

        // iterate all possible start point(0~wordLen-1)
        for (int i = 0; i < wordLen; i++) {
            int left = i, right = i;
            Map<String, Integer> curCount = new HashMap<>();

            // expand right window
            while (right + wordLen <= s.length()) {
                String w = s.substring(right, right + wordLen);
                right += wordLen;

                // if cur word in words
                if (map.containsKey(w)) {
                    curCount.put(w, curCount.getOrDefault(w, 0) + 1);

                    // if cur word exceed the num in words -> shrink window
                    while (curCount.get(w) > map.get(w)) {
                        String leftWord = s.substring(left, left + wordLen);
                        curCount.put(leftWord, curCount.get(leftWord) - 1);
                        left += wordLen;
                    }

                    // if window size == target size -> valid
                    if (right - left == windowSize) {
                        res.add(left);
                        String leftWord = s.substring(left, left + wordLen);
                        curCount.put(leftWord, curCount.get(leftWord) - 1);
                        left += wordLen;
                    }
                } else {
                    // if cur word not in words, clear cur window
                    curCount.clear();
                    left = right;
                }
            }
        }
        return res;
    }
}

File: 0001-0500/0030. Substring with Concatenation of All Words.py
# Time:O(n*m), Space:O(k)

from typing import List, Counter

class Solution:
    def findSubstring(self, s: str, words: List[str]) -> List[int]:
        if not s or not words:
            return []

        word_len = len(words[0])
        word_count = len(words)
        total_len = word_len * word_count
        word_freq = Counter(words)
        res = []

        for i in range(word_len):
            left, right = i, i
            cur_freq = Counter()

            while right + word_len <= len(s):
                word = s[right : right + word_len]
                right += word_len

                if word in word_freq:
                    cur_freq[word] += 1

                    while cur_freq[word] > word_freq[word]:
                        left_word = s[left : left + word_len]
                        cur_freq[left_word] -= 1
                        left += word_len

                    if right - left == total_len:
                        res.append(left)
                else:
                    cur_freq.clear()
                    left = right

        return res
